RP1 Video Out: Hack for 50fps/60fps buffer-flips in interlaced modes

njhollinghurst opened the pull request:
By avoiding VEC's built-in interlaced DMA modes, and sending individual fields (adapting the "field wobble" trick from the DPI driver), it looks like we can get VEC to update at full field rate.

This changes the pattern of VSync pulses on the VEC's FE->BE interface. Some jiggery-pokery with the NTSC vertical settings was required before the BE would tolerate this (my guess is that the offset between "frame datum" and VSync, which is nonzero in conventional NTSC line numbering, made it more sensitive to FE VSync timings?)

Needs more testing to ensure it's reliable in all modes (including the "vintage" ones). Unlike the DPI case we can't use PIO to infer correct field polarity. If it  ever gets out of step, it won't recover!

To clarify: The first of these two commits relates to _progressive_ modes, and is unrelated.

pelwell commented on drivers/gpu/drm/rp1/rp1-vec/rp1_vec_hw.c:
```diff
@@ -525,32 +574,48 @@ void rp1vec_hw_setup(struct rp1_vec *vec,
 {
+	unsigned long flags;
+
+	spin_lock_irqsave(&vec->hw_lock, flags);
```
Could the spin_lock be claimed later (more for readability than efficiency)?

pelwell commented on drivers/gpu/drm/rp1/rp1-vec/rp1_vec_hw.c:
```diff
@@ -525,32 +574,48 @@ void rp1vec_hw_setup(struct rp1_vec *vec,
-	if (vec->fake_31khz) {
-		a += stride;
+	addr += offset;
+	vec->last_dma_addr = addr;
```
Here, perhaps?

njhollinghurst replied:
It could skip one line `addr += offset;` but it's then needed to protect the atomicity of `last_dma_addr` and `last_stride` as well as the register writes.

njhollinghurst replied:
Moved

njhollinghurst committed:
drm: rp1: rp1-vec: Support 60fps in interlaced modes; other tweaks

To work around the 30fps buffer-flip rate limit when using VEC's
"native" interlaced modes, switch to sending individual fields
to the VEC BE, using an ISR to flip between fields.

When the TV mode is NTSC, change advertised progressive modes to
have 263 total lines; this ameliorates colour artifacts, although
it reduces the frame rate slightly from 60.05Hz to 59.83Hz.
Progressive modes with 262 lines remain supported.

Fix an error in equalising pulse configuration for PAL-M/PAL60.

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>
